hcl/hcl.go

```diff
@@ -208,13 +212,92 @@ func (p *TerramateParser) AddFile(name string, data []byte) error {
 
 // Parse the previously added files and return either a Config or an error.
 func (p *TerramateParser) Parse() (Config, error) {
-	err := p.parseSyntax()
+	err := p.MinimalParse()
 	if err != nil {
 		return Config{}, err
 	}
+
+	// TODO(i4k): don't validate schema here.
+	// Changing this requires changes to the editor extensions / linters / etc.
 	return p.parseTerramateSchema()
 }
 
+// MinimalParse does the syntax parsing and merging of configurations but do not
+// validate if it's valid terramate configuration.
+func (p *TerramateParser) MinimalParse() error {
+	err := p.parseSyntax()
+	if err != nil {
+		return err
+	}
+
+	return p.mergeConfig()
+}
+
+func (p *TerramateParser) mergeConfig() error {
+	errs := errors.L()
+
+	unmerged := func(origin string, block *hclsyntax.Block) {
+		p.Blocks = append(p.Blocks, NewBlock(origin, block))
+	}
+
+	merged := func(origin string, block *hclsyntax.Block) {
+		if other, ok := p.MergedBlocks[block.Type]; ok {
+			errs.Append(other.MergeBlock(origin, block))
+		} else {
+			merged := NewMergedBlock(block.Type)
+			p.MergedBlocks[block.Type] = merged
+			errs.Append(merged.MergeBlock(origin, block))
+		}
+	}
+
+	type mergeHandler func(origin string, block *hclsyntax.Block)
+	mergeHandlers := map[string]mergeHandler{
+		"terramate": merged,
+		"globals":   merged,
+
+		"stack":         unmerged,
+		"generate_file": unmerged,
+		"generate_hcl":  unmerged,
+	}
+
+	fileMap := p.ParsedFiles()
+	for _, origin := range p.sortedFilenames() {
+		body := fileMap[origin]
```

Comment:
```suggestion
	bodiesMap := p.ParsedFiles()
	for _, origin := range p.sortedFilenames() {
		body := bodiesMap[origin]
```

Because of fileMap I thought body was the contents of a file, not a parsed body.